### PR TITLE
[tflite] add fp32->fp16 relaxation option to NNAPI delegate

### PR DESCRIPTION
**Pull request by freedomtan**
Most NN accelerators on Android devices don't support FP32. To evaluation their capabilities, we should call `Interpreter::SetAllowFp16PrecisionForFp32(True)` so that NNAPI delegate will call the corresponding [ANeuralNetworksModel_relaxComputationFloat32toFloat16()](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1gab822719f98f0c92e5da3684cdaca6ba0) when used.

This patch adds `fp16` related support so that we can do something like
```
./imagenet_accuracy_eval \
--model_file=tflite-models-new/float/mobilenet_v1_1.0_224.tflite \
--ground_truth_images_path=ilsvrc_images \
--ground_truth_labels=ilsvrc_validation_labels.txt \
--model_output_labels=labels.txt \
--output_file_path=accuracy_outputs/mobilenet_v1_1.0_224_float_nnapi.txt \
--delegate=nnapi \
--allow_fp16=1 \
--num_images=0
```
on Android devices

### CODE REVIEW

**Comment by teijeong** [comment]
Hi Chao, can you look at this one?

**Comment by multiverse-tf** [comment]
> Hi Chao, can you look at this one?

Sure.

**Review by multiverse-tf — COMMENTED** [review]
At a high level, this particular interpreter API (i.e. SetAllowFp16PrecisionForFp32) is tied to the legacy NNAPI delegate implementation and it's experimental (see the comment of the API). So, I prefer not to achieve this feature via these changes, but preferring a change via adding an option to NNAPI delegate implementation.

For example, the GPU delegate also supports FP16, but it's enabled by specifying the delegate option (i.e. inference_priority1/2/3) when creating the GPU delegate.

Note that NNAPI delegate has an option called execution_preference (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/nnapi/nnapi_delegate.h#L55). Could you use such options to meet your requirements? Particularly, when using the evaluation or the benchmark tool, you could add "--nnapi_execution_preference=<..>" flag to enable this together with "--use_nnapi=true".  See https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/delegates#nnapi-delegate-provider for details. The implementation could be referred to the lite/tools/delegates/nnapi_delegate_provider.cc file.

Btw,  FP32->FP16 relaxation for the GPU delegate is enabled by default at the moment when using the evaluation/benchmark tool, This could be turned off by "--gpu_precision_loss_allowed=false" flag. See lite/tools/delegates/gpu_delegate_provider.cc for details.

**Comment by freedomtan** [comment]
@multiverse-tf Agree that `SetAllowFp16PrecisionForFp32()` may not be great, but it seems `execution_preference`, which I think should stick to what NNAPI has, is not the right place to add the fp16 option (I submitted Java wrapper to the `execution_preference` before, 996f0ec). I can add a fp16 field after [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/nnapi/nnapi_delegate.h#L155-L186) to replace the Interpreter option, but that means more places should be changed, including benchmark_model, evaluation code, Java wrapper, etc.

**Comment by multiverse-tf** [comment]
> @multiverse-tf Agree that `SetAllowFp16PrecisionForFp32()` may not be great, but it seems `execution_preference`, which I think should stick to what NNAPI has, is not the right place to add the fp16 option (I submitted Java wrapper to the `execution_preference` before, [996f0ec](https://github.com/tensorflow/tensorflow/commit/996f0ec8b3e915a3264b409ff92011465dd8647b)). I can add a fp16 field after [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/nnapi/nnapi_delegate.h#L155-L186) to replace the Interpreter option, but that means more places should be changed, including benchmark_model, evaluation code, Java wrapper, etc.

TBH, I think adding this fp32->f16 relaxation option to nnapi delegate is simpler and less invasive than this one. In particular, as for tooling side change, you just need to add the new option to nnapi_delegate_provider.cc  to enable the support across evaluation/benchmark tools w/o changing evaluation-inference-stage internals. From this perspective, the approach is more general, benefiting the benchmark tool as well as improving the nnapi delegate implementation itself.

In addition, I think it's straightforward to change this in separate smaller PRs (or in a single PR if you want) :
1. add the option to C++ nnapi delegate implementation
Basically, one could pass the 'delegate_options' retrieved in NNAPIDelegateKernel::Init to NNAPIDelegateKernel::BuildGraph where context->allow_fp32_relax_to_fp16 is used, and then use the new option.
2. add the option to the Java wrapper.
3. support the option on the tooling side via nnapi_delegate_provider.

I'm happy to review the PRs.

**Comment by freedomtan** [comment]
rebased, added an option to NNAPI delegate, changed evaluation tools to use the newly added option by `--nnapi_allow_fp16` , and added it to `benchmark_model`.

I'll send Java related changes in another PR because I don't have good Java development environment to test them now.

File: tensorflow/lite/delegates/nnapi/nnapi_delegate.cc [hunk 1]
```diff
@@ -3202,6 +3203,7 @@ TfLiteStatus NNAPIDelegateKernel::Prepare(TfLiteContext* context,
 
   const auto delegate_options =
       StatefulNnApiDelegate::GetOptions(node->delegate);
+
```
Comment (multiverse-tf) [inline tensorflow/lite/delegates/nnapi/nnapi_delegate.cc:3206]: Remove this extra line?

File: tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc [hunk 2]
```diff
@@ -141,6 +141,8 @@ class CompositeObserver : public ImagenetModelEvaluator::Observer {
       tflite::Flag::CreateFlag(kNumRanksFlag, &params.num_ranks,
                                "Generates the top-1 to top-k accuracy values"
                                "where k = num_ranks. Default: 10"),
+      tflite::Flag::CreateFlag("nnapi_allow_fp16", &params.nnapi_allow_fp16,
+                               "allow fp16 in nnapi"),
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc:145]: I think we don't need to change this code as "tools/evaluation/tasks/imagenet_image_classification" is a better version of this one. We should have done some more proactive legacy code clean up.

File: tensorflow/lite/tools/delegates/default_execution_provider.cc [hunk 3]
```diff
@@ -30,6 +30,7 @@ class DefaultExecutionProvider : public DelegateProvider {
                              ToolParam::Create<int32_t>(0));
     default_params_.AddParam("min_nodes_per_partition",
                              ToolParam::Create<int32_t>(0));
+    default_params_.AddParam("allow_fp16", ToolParam::Create<bool>(false));
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/delegates/default_execution_provider.cc:33]: These changes aren't needed as we limit the new option to nnapi delegate?

File: tensorflow/lite/tools/benchmark/benchmark_performance_options.cc [hunk 4]
```diff
@@ -302,6 +303,9 @@ void BenchmarkPerformanceOptions::CreatePerformanceOptions() {
                         BenchmarkParam::Create<bool>(false));
         params.AddParam("max_delegated_partitions",
                         BenchmarkParam::Create<int>(0));
+        params.AddParam("max_delegated_partitions",
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/benchmark/benchmark_performance_options.cc:306]: These 3 lines aren't needed.

File: tensorflow/lite/tools/evaluation/evaluation_delegate_provider.cc [hunk 5]
```diff
@@ -132,6 +132,10 @@ tools::ToolParams DelegateProviders::GetAllParams(
     tool_params.Set<int32_t>("num_threads", params.num_threads());
   }
 
+  if (params.has_nnapi_allow_fp16()) {
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/evaluation/evaluation_delegate_provider.cc:135]: These changes are not needed as the flag is already added to the nnapi_delegate_provider.cc

File: tensorflow/lite/tools/evaluation/tasks/coco_object_detection/run_eval.cc [hunk 6]
```diff
@@ -65,6 +65,7 @@ class CocoObjectDetection : public TaskExecutor {
   bool debug_mode_;
   std::string delegate_;
   int num_interpreter_threads_;
+  bool allow_fp16_;
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/evaluation/tasks/coco_object_detection/run_eval.cc:68]: This option will be parsed and accessible via delegate_providers_. So, the change isn't needed.

File: tensorflow/lite/tools/evaluation/tasks/coco_object_detection/run_eval.cc [hunk 7]
```diff
@@ -104,6 +105,9 @@ CocoObjectDetection::CocoObjectDetection(int* argc, char* argv[])
           kDelegateFlag, &delegate_,
           "Delegate to use for inference, if available. "
           "Must be one of {'nnapi', 'gpu', 'xnnpack', 'hexagon'}"),
+      tflite::Flag::CreateFlag(
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/evaluation/tasks/coco_object_detection/run_eval.cc:108]: Same comment w/ the previous one. Similarly, all changes to run_eval.cc in the following are not needed.

File: tensorflow/lite/delegates/nnapi/nnapi_delegate.cc [hunk 8]
```diff
@@ -4021,6 +4027,7 @@ StatefulNnApiDelegate::StatefulNnApiDelegate(const NnApi* nnapi,
       options.max_number_delegated_partitions;
   TFLITE_LOG_PROD_ONCE(tflite::TFLITE_LOG_INFO,
                        "Created TensorFlow Lite delegate for NNAPI.");
+  delegate_data_.allow_fp16 = options.allow_fp16;
```
Comment (multiverse-tf) [inline tensorflow/lite/delegates/nnapi/nnapi_delegate.cc:4030]: Move this before the TFLITE_LOG_PROD_ONCE?

File: tensorflow/lite/delegates/nnapi/nnapi_delegate.cc [hunk 9]
```diff
@@ -3885,6 +3888,7 @@ TfLiteStatus NNAPIDelegateKernel::BuildGraph(
   std::vector<uint32_t> outputs;
   outputs.reserve(output_tensors->size);
 
+  const auto delegate_options = StatefulNnApiDelegate::GetOptions(delegate);
```
Comment (multiverse-tf) [inline tensorflow/lite/delegates/nnapi/nnapi_delegate.cc:3891]: As this is a copy and the delegate_options is already obtained in the caller, what about just passing a const reference to StatefulNnApiDelegate::Options instead of 'delegate' pointer?

File: tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc (hunk 2 above)
Comment (freedomtan) [reply tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc:145]: I like the output of this one (top-1, .., top-10 for each image)
```
...
67.742, 83.871, 87.097, 90.323, 90.323, 90.323, 93.548, 93.548, 93.548, 93.548
68.750, 84.375, 87.500, 90.625, 90.625, 90.625, 93.750, 93.750, 93.750, 93.750
69.697, 84.848, 87.879, 90.909, 90.909, 90.909, 93.939, 93.939, 93.939, 93.939
70.588, 85.294, 88.235, 91.176, 91.176, 91.176, 94.118, 94.118, 94.118, 94.118
...
```

File: tensorflow/lite/delegates/nnapi/nnapi_delegate_kernel.h [hunk 10]
```diff
@@ -349,7 +349,8 @@ class NNAPIDelegateKernel {
   TfLiteStatus AddOpsAndTensors(TfLiteContext* context, int* nnapi_errno);
 
   TfLiteStatus BuildGraph(TfLiteContext* context,
-                          const TfLiteIntArray* input_tensors,
+                          const StatefulNnApiDelegate::Options& options,
+			  const TfLiteIntArray* input_tensors,
```
Comment (multiverse-tf) [inline tensorflow/lite/delegates/nnapi/nnapi_delegate_kernel.h:353]: bad indent?

File: tensorflow/lite/tools/evaluation/proto/evaluation_stages.proto [hunk 11]
```diff
@@ -121,6 +121,9 @@ message TfliteInferenceParams {
   // This helps benchmark cases where extensive pre-processing might not be
   // required for every input.
   optional int32 invocations_per_run = 4 [default = 1];
+
+  // nnapi_allow_fp16
+  optional bool nnapi_allow_fp16 = 5 [default = false];
```
Comment (multiverse-tf) [inline tensorflow/lite/tools/evaluation/proto/evaluation_stages.proto:126]: Drop this one? To support using a new field in this proto, I think more changes are needed. In this case, as the option is enabled by delegate_provider, we don't need this new field.

File: tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc (hunk 2 above)
Comment (multiverse-tf) [reply tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc]: Acked. I checked the code again, and found this tool actually supports the delegate registrar mechanism, i.e. the new option you added will be automatically supported w/ this tool :-)

File: tensorflow/lite/delegates/nnapi/nnapi_delegate_kernel.h (hunk 10 above)
Comment (freedomtan) [reply tensorflow/lite/delegates/nnapi/nnapi_delegate_kernel.h:353]: tab :-(, my bad, didn't run `clang-format` on .h

File: tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc (hunk 2 above)
Comment (freedomtan) [reply tensorflow/lite/tools/accuracy/ilsvrc/imagenet_model_evaluator.cc]: Yes, it works. I tried. Thanks

File: tensorflow/lite/tools/evaluation/proto/evaluation_stages.proto (hunk 11 above)
Comment (freedomtan) [reply tensorflow/lite/tools/evaluation/proto/evaluation_stages.proto:126]: yes, forget to delete this. Thanks

**Review by multiverse-tf — APPROVED** [review]
